### PR TITLE
Initial working build

### PR DESCRIPTION
**Pull request by lmvdz**
Important note:

Don't use the same account for the bot as you use for the HubbleProtocol (UI)
When you borrow with SOL as collateral it closes the WSOL account, so you won't be able to claim gains if you use the same wallet.

### CODE REVIEW

File: .prettierrc.js [hunk 1]
```diff
@@ -0,0 +1,9 @@
+module.exports = {
```
Comment (y2kappa) [inline .prettierrc.js:1]: nitpick: add a linting rule in github actions to follow this

File: README.md [hunk 2]
```diff
@@ -1 +1,15 @@
-# liq-bot
+# Typescript Liquidation Bot for HubbleProtocol by Lmvdzande
```
Comment (y2kappa) [inline README.md:1]: Some explanation about why we need both tryLiquidate and clearLiquidation gains could help

File: src/types.ts [hunk 3]
```diff
@@ -0,0 +1,422 @@
...
+
+export type LiquidatorAssociatedTokenAccounts = {
+    [key in CollateralTokenActive]: string;
+};
+
+export type Loan = {
+    metadata: UserMetadata,
+    usdhDebt: BN | Decimal
+    collateral: CollateralAmounts,
+    tcr: BN | Decimal
```
Comment (y2kappa) [inline src/types.ts]: tcr stands for TotalCollateralRatio and refers to the entire System, whereas for a user it's typically ICR (individual collateral ratio)

File: src/liquidator.ts [hunk 4]
```diff
@@ -0,0 +1,647 @@
...
+import { PollingAccountsFetcher, flat, chunk } from 'polling-account-fetcher';
+import { TpuConnection } from 'tpu-client';
+import Decimal from "decimal.js";
+import * as bs58 from 'bs58';
+
+import { config } from 'dotenv';
+config();
+
+function getWallet(): Wallet {
+	const botKeyEnvVariable = "BOT_KEY";
```
Comment (y2kappa) [inline src/liquidator.ts]: prefer 2 spaces instead of a 4-spaces wide tab

File: src/liquidator.ts [hunk 5]
```diff
@@ -0,0 +1,647 @@
...
+import { Config, BorrowingMarketState, LiquidationsQueue, LiquidatorAssociatedTokenAccounts, UserMetadata, TokenPrices, CollateralTokenActive, CollateralAmounts, TokenMap, Loan, CollateralInfo, SystemMode } from "./types";
+import { PollingAccountsFetcher, flat, chunk } from 'polling-account-fetcher';
+import { TpuConnection } from 'tpu-client';
+import Decimal from "decimal.js";
+import * as bs58 from 'bs58';
+
+import { config } from 'dotenv';
+config();
+
+function getWallet(): Wallet {
```
Comment (y2kappa) [inline src/liquidator.ts]: would move this into a utils.ts

File: src/liquidator.ts [hunk 6]
```diff
@@ -0,0 +1,647 @@
...
+                }
+            }, (error) => {
+                console.error(error);
+            });
+        });
+    }
+
+    // create the clearLiquidationGains TransactionInstruction for the liquidator running this bot
+    // will need to be called for each token
+    async getClearLiquidationGainsIX(token: string, clearerAassociatedTokenAccount: PublicKey) : Promise<TransactionInstruction> {
```
Comment (y2kappa) [inline src/liquidator.ts]: you could also generate an anchor Provider client that already has the rpc endpoints built in

File: src/liquidator.ts [hunk 7]
```diff
@@ -0,0 +1,647 @@
...
+            tx.add(this.clearLiquidationGainsIXMap.get(token));
+            tx.feePayer = this.hubbleProgram.provider.wallet.publicKey;
+            tx.recentBlockhash = blockhash;
+            tx = await this.hubbleProgram.provider.wallet.signTransaction(tx);
+            return await (this.hubbleProgram.provider.connection as TpuConnection).sendRawTransaction(tx.serialize());
+        })));
+    }
+
+    // get the pending debt of a user
+    async getPendingDebt(user : UserMetadata) : Promise<Decimal> {
```
Comment (y2kappa) [inline src/liquidator.ts]: any reason you are not using our sdk and recreating this?

File: src/index.ts [hunk 8]
```diff
@@ -0,0 +1,8 @@
+import { Bot } from "./bot";
+
+async function main() {
+    const bot = await Bot.create();
+    bot.start();
```
Comment (y2kappa) [inline src/index.ts]: await?

**Review by y2kappa — COMMENTED** [review]
good stuff, a few general suggestions, nothing major

File: src/index.ts (hunk 8 above)
Comment (lmvdz) [reply src/index.ts]: The create function is async, does calls to get the idl and config, much cleaner to read by using async/await

File: src/liquidator.ts (hunk 7 above)
Comment (lmvdz) [reply src/liquidator.ts]: Initially I had just used BN then realized the price was getting it's decimals truncated, then I added the Decimal... I'd like to move to using the sdk!

File: .prettierrc.js (hunk 1 above)
Comment (lmvdz) [reply .prettierrc.js:1]: I'll need to figure that one out

File: README.md (hunk 2 above)
Comment (lmvdz) [reply README.md:1]: Can do!

File: src/types.ts (hunk 3 above)
Comment (lmvdz) [reply src/types.ts]: Okay will make that distinction

File: src/liquidator.ts (hunk 7 above)
Comment (lmvdz) [reply src/liquidator.ts]: It works better this way, the UserMetadata is stored on the polling-accounts-fetcher accounts map. I'd rather not have duplicate data on the bot (UserMetadata BN) and (UserMetadata Decimal)

**Pull request merged**